service/timer_event_source.cc

```diff
@@ -0,0 +1,193 @@
+    }
+}
+
+TimerEventSource::
+~TimerEventSource()
+{
+    int res = ::close(timerFd_);
+    if (res == -1) {
+        cerr << "warning: close on timerfd: " << strerror(errno) << endl;
+    }
+}
+
+int
+TimerEventSource::
+selectFd()
+    const
+{
+    return timerFd_;
+}
+
+bool
+TimerEventSource::
+processOne()
+{
+start:
+    uint64_t numWakeups = 0;
+    int res = ::read(timerFd_, &numWakeups, 8);
+    if (res == -1) {
+        if (errno == EINTR) {
+            goto start;
+        }
+        else if (errno == EAGAIN || errno == EWOULDBLOCK) {
+            return false;
+        }
+        else {
+            throw ML::Exception(errno, "timerfd read");
+        }
+    }
+    ExcAssertEqual(res, 8);
+    onTimerTick();
+
+    return false;
+}
+
+void
+TimerEventSource::
+onTimerTick()
+{
+    Date now = Date::now();
+    vector<Timer> triggered = collectTriggeredTimers(now);
+
+    for (auto & timer: triggered) {
+        double interval = timer.lastTick.secondsUntil(now);
+        uint64_t numTicks = (uint64_t) ::floor(interval / timer.interval);
+        if (timer.onTick(numTicks)) {
+            timer.lastTick = now;
+            timer.nextTick = now.plusSeconds(timer.interval);
+            insertTimer(move(timer));
+        }
+    }
+
+    adjustNextTick(now);
+}
+
+vector<TimerEventSource::Timer>
+TimerEventSource::
+collectTriggeredTimers(Date refDate)
+{
+    vector<Timer> triggered;
+    TimersGuard guard(timersLock_);
+
+    size_t nbrTriggered, nbrTimers(timerQueue_.size());
+    for (nbrTriggered = 0; nbrTriggered < nbrTimers; nbrTriggered++) {
+        if (timerQueue_[nbrTriggered].nextTick > refDate) {
+            break;
+        }
+    }
+
+    if (nbrTriggered > 0) {
+        triggered.reserve(nbrTriggered);
+        for (size_t i = 0; i < nbrTriggered; i++) {
+            triggered.emplace_back(move(timerQueue_[i]));
+        }
+        timerQueue_.erase(timerQueue_.begin(), timerQueue_.begin() + nbrTriggered);
+    }
+
+    return triggered;
+}
+
+void
+TimerEventSource::
+adjustNextTick(Date now)
+{
+    TimersGuard guard(timersLock_);
+
+    Date negInfinity = Date::negativeInfinity();
+    Date nextTick = ((timerQueue_.size() > 0)
+                     ? timerQueue_[0].nextTick
+                     : negInfinity);
+
+    if (nextTick != nextTick_) {
+        struct itimerspec spec{{0, 0}, {0, 0}};
+
+        if (nextTick != negInfinity) {
+            double delta = nextTick - now;
+            auto & value = spec.it_value;
+            value.tv_sec = (time_t) delta;
+            value.tv_nsec = (delta - spec.it_value.tv_sec) * 1000000000;
+        }
+        int res = ::timerfd_settime(timerFd_, 0, &spec, nullptr);
+        if (res == -1) {
+            throw ML::Exception(errno, "timerfd_settime");
+        }
+        nextTick_ = nextTick;
+    }
+}
+
+uint64_t
+TimerEventSource::
+addTimer(double delay, const OnTick & onTick)
+{
+    Date now = Date::now();
+    uint64_t timerId = counter_.fetch_add(1);
```

Comment:
@jeremybarnes indeed and this raises a design question that I worked-around by using an assertion... I do not see how addTimer can guarantee that the ids will be unique, other than by remembering them somehow. I considered the use of a guid instead, but even though it postpones it further in time, it does not solve the problem either.